Design.md

```diff
@@ -0,0 +1,107 @@
+* <rep_policy> := RS,k,m or CP,n (for n-1 copies)
+* <object_key> := parent S3 object key (or a prefix) - can contain anything but ‘#’
+* <rand> := 64 bits random number (unicity inside 1 hyperdrive)
+* <location>:= hyperdrive location (UUID, idx in table?, ip:port, ...)
+
+The fragments do not store the exact, really used hyperdrive keys. Those can be derived easily with only the generated key, even for split.
+* <stored_fragment_key> := <object_key>-<rand>-<start_offset>-<type>-<fragid>
+* <object_key> and <rand> are the ones defined above
+* <type> := ‘d’ for data, ‘c’ for coding
+* <fragid>:= index in main key fragment list
+* <start_offset> := used for splits. All split chunks share the same prefix, storing the offset is used to easily have range queries and avoid storing them all in the main key.
+
+Meets all requirements (except perhaps around the last stripe size of an ECN chunk). Split is required to contact the same hyperdrives for the object. This has several benefits: only selecting once, object located on only a handful of hyperdrives => less machines to contact = less opportunities to fail or hit a straggler. The way split is handled also enables us not to have a manifest, worries about its size, location or freshness.
+
+Example key:
+1/ Small key: storing s3://fakebucket/obj1 with RS2+1 onto hd1 hd2 and hd3
+Main key: 1#1#1,0#RS,2,1#obj1#314159#hd1#hd3#hd2
+Hyperdrive keys:
+1. obj1-314159-0-d-1
+2. obj1-314159-0-d-2
+3. obj1-314159-0-c-3
+
+2/ Splitted key: storing s3://fakebucket/Large1 with RS2+1 onto hd1 hd2 and hd3
+Key size: 64000, split_size: 49000, 2 parts on same hyperdrive (no conflict!)
+
+Main key: 1#1#2,49000#RS,2,1#Large1#42#hd3#hd2#hd3
+Hyperdrive keys:
+* On hd1: none
+* On hd2
+  1. Large1-42-0-d-2
+  2. Large1-42-49000-d-2
+* On hd3
+  1. Large1-42-0-d-1
+  2. Large1-42-0-c-3
+  3. Large1-42-49000-d-1
+  4. Large1-42-49000-c-3
+
+## Error handling
+
+### On DELETE
+
+Failure to delete a fragment ( != 20X, 404) should not fail the overall operation but must log an appropriate entry into designated Kafka topic, to be cleaned later on by HdRepair.
+
+Topic: orphans
+Content: JSON messages
+Example JSON:
+```json
+{
+    "objectKey": ... # S3 object key
+    "rawKey": .... # associated hdclient metadata
+    "orphans": [
+        hyperdrive_key_missing_fragment1,
+        hyperdrive_key_missing_fragment2,
+        hyperdrive_key_missing_fragment3,
+        ...
+    ]
+}
+```
+
+### On PUT
+
+Failure ot pUT any fragment (erasure coded, replicated or split) should fail the overall operation. All fragments already PUT pending must be cleaned afterwards. Cleaning is handled in the same way as orphans created on deletion (refer to On DELETE section).
+
+Special case of timeout/connection close: a timeout on a PUTing a fragment leaves no clear signal. Is it stored? Queued but not yet written? We could either:
+1. Consider it as not OK and fail the overall PUT
```

Comment:
my 2 cents on this problem: There are multiple corner-cases here:
1. The operation took time to complete but was actually completed, and the HyperDrive did not have time to answer
 -> OK seen as ETIMEOUT (considering it a failure requires cancelling/rolling-back the operation, which may be complex)
2. The operation somehow got stuck and did not complete
 -> NOK seen as ETIMEOUT (With proper handling, the request can thus be cancelled internally, without a need to clean anything up beyond that cancellation)

Given these two cases, in the past the MD team has always considered that such a case should:
1. Cancel the operation if possible
2. Make it so that the system isn't "dirty", as much as possible (ie: have a completed/valid operation, or making it so that the operation never happened if possible)
3. Leave the status checking to the "client", not taking any further decision on the data itself (Note that this is also true for the S3 layer, a TIMEOUT means one needs to check the status of the oepration through observation of the new state, so we kind of extended it to underlaying systems)

Reply:
'Note that this is also true for the S3 layer' I didn't know that when writing those specs. Interesting

Reply:
Won't be fixed right now. What exact action to do on what scenario is pretty much still under discussion. Implementation, however should be entirely flexible: ie having a kind of context/global view of the whole operation.

Note: it is not the case in this branch. I have a single request, so seemed over engineered for a quick demoable v1